### ISSUE TITLE
Обучение модели

### ISSUE THREAD

**Issue by k0bozeva** [Kulik1987/Profit-pulse-docs#11]
Изучение ресурсов https://www.kaggle.com/ и https://huggingface.co/models
Поиск на https://www.kaggle.com/ и https://huggingface.co/models велся по следующим ключевым словам: “mental health conversations” и “psychology”. Приемлемые варианты находились по первому ключевому словосочетанию.

Какие модели могут подойти нам для применения? Есть ли предобученные на данных по психологии модели (как вариант попробовать в название psychology поискать)?
На kaggle есть несколько обученных моделей, подходящих под бизнес-цель:
[A WellBot](https://www.kaggle.com/models/chidinmaikhinmwin/bert) - основана на архитектуре GPT-2;
[chappie](https://www.kaggle.com/models/georgecane/chappie) - основана на LNBM - модели обработки естественного языка (NLP), такие как BERT, GPT2 и flan-t5-small, со сверточными нейронными сетями (CNNS) и архитектурами ResNet (остаточных сетей);
[MedGenius_LLaMA-3.2B](https://www.kaggle.com/models/huzefanalkheda/medgenius_llama-3.2bv1) - модель искусственного интеллекта, созданная на основе набора данных Medical Intelligence, состоящего из более чем 40 000 подробных строк медицинской информации;
На huggingface моделей представлено больше, для первичного ознакомления необходимо время
Посмотреть датасеты для дообучения psychology что интересует: нам нужны датасеты где не ставятся диагнозы и даются развернутые ответы, не одним предложением. Так же посмотреть какие структуры у датасетов и разобраться в чем эти структуры отличаются и на что влияет различные структуры в датасетах.
датасеты на [kaggle](https://www.kaggle.com/search?q=mental+health+conversations+sortBy%3Adate+in%3Adatasets) - нашлось 28 штук. Выбрал наиболее релевантные. В целом, в них представлены подходящие ответы, например:
Датасет [Mental_Health_Conversations_Dataset](https://www.kaggle.com/datasets/chidinmaikhinmwin/mental-health-conversations-dataset) представлен набором Запрос - ответ. Формат файла json. 3500 пар вопрос/ответ. Диагнозы, как таковые не ставятся, информация представлена в виде:
Запрос: “Я переживаю некоторые моменты, связанные со своими чувствами и с самим собой. Я почти не сплю и ничего не делаю, только думаю о том, какой я никчемный и что меня здесь не должно быть. Я никогда не пытался покончить с собой и не помышлял о нем. Я всегда хотел решить свои проблемы, но у меня никак не доходили до этого руки. Как я могу избавиться от ощущения, что я никому не нужен?”
Ответ: “Если все считают вас никчемным, то, возможно, вам нужно найти новых людей, с которыми вы могли бы общаться. Серьезно, социальный контекст, в котором живет человек, оказывает большое влияние на его самооценку. В противном случае, вы можете ходить по кругу, пытаясь понять, почему вы не бесполезны, а затем вернуться в ту же толпу и снова быть сбитым с толку. В социальных сетях вы можете найти множество вдохновляющих сообщений. Возможно, вы прочтете некоторые из них, в которых утверждается, что нет никчемных людей и что у каждого есть благая цель в жизни. Кроме того, поскольку наша культура настолько пропитана верой в то, что если кто-то недоволен собой, это в некотором роде ужасно. Плохие чувства - это часть жизни. Они побуждают нас избегать ситуаций и отношений, которые приносят нам больше вреда, чем пользы. Плохие чувства действительно ужасны. Ваше чувство никчемности может быть полезным в том смысле, что оно побуждает вас осознать, что вы намного лучше, чем ваши сегодняшние чувства.”
Датасет [Synthetic_Mental_Health_Conversations](https://www.kaggle.com/datasets/denisemtatih/synthetic-mental-health-conversations) представлен набором Запрос - ответ. Формат файла csv. Свыше 300 тысяч пар вопрос/ответов.
Датасет [Mental Health Counseling Conversations](https://www.kaggle.com/datasets/melissamonfared/mental-health-counseling-conversations-k) представлен набором Запрос - ответ. Формат файла json.
Датасет [Human and LLM Mental Health Conversations](https://www.kaggle.com/datasets/birdy654/human-and-llm-mental-health-conversations) представлен набором Запрос - ответ. Формат файла csv.
Датасет [Amod Mental Health Counseling Conversations](https://www.kaggle.com/datasets/thedevastator/amod-mental-health-counseling-conversations-data) представлен набором Запрос - ответ. Формат файла csv.
Датасет [ProsocialDialog - Problematic Content Dialogue](https://www.kaggle.com/datasets/thedevastator/prosocialdialog-dialogue-dataset) представлен набором Запрос - ответ. Формат файла csv.
Есть также датасеты, в которых оценивается состояние, проводится некая категоризация клиента, есть датасеты для постороения непрерывного диалога, например [Mental Health Conversational Data](https://www.kaggle.com/datasets/elvis23/mental-health-conversational-data) – в нем 3 столбца: “tag”, “patterns”, “responses”:
“tag”: string"neutral-response"
“patterns”:[string"nothing much"]
“responses”:**[**string"Oh I see. Do you want to talk about something?" ]
датасеты на [huggingface](https://huggingface.co/):
Датасет [tcabanski/mental_health_counseling_conversations_rated](https://huggingface.co/datasets/tcabanski/mental_health_counseling_conversations_rated) представлен набором Запрос - ответ. Формат файла json.
Датасет [zaghloul2012/mental-health-conversations](https://huggingface.co/datasets/zaghloul2012/mental-health-conversations) представлен набором Запрос - ответ. Формат файла json, 3 выборки
1. Тестирование качества ответов модели
Тестирование модели может включать различные аспекты.

a. Метрики оценки
Для оценки качества ответов модели можно использовать следующие метрики:

Accuracy
Precision, Recall, F1 Score
BLEU / ROUGE
Perplexity
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Пример: истинные метки и предсказания модели
true_answers = ["Да", "Нет", "Да", "Да", "Нет"]
predicted_answers = ["Да", "Нет", "Нет", "Да", "Нет"]

# Расчет метрик
accuracy = accuracy_score(true_answers, predicted_answers)
precision = precision_score(true_answers, predicted_answers, pos_label="Да", average="binary")
recall = recall_score(true_answers, predicted_answers, pos_label="Да", average="binary")
f1 = f1_score(true_answers, predicted_answers, pos_label="Да", average="binary")

print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print(f"F1 Score: {f1:.2f}")
BLEU (Bilingual Evaluation Understudy) и ROUGE (Recall-Oriented Understudy for Gisting Evaluation) — это популярные метрики для оценки качества генерируемого текста, особенно в задачах машинного перевода и обработки естественного языка, таких как резюмирование.
BLEU Score
BLEU измеряет соответствие между генерируемым текстом и одним или несколькими эталонными ответами. Оценка варьируется от 0 до 1, где 1 указывает на выставление абсолютно идентичных текстов. Основные этапы расчета BLEU включают:

N-граммы: Сравнение N-грамм (последовательностей из N слов) генерируемого текста и эталонных ответов.
Precision: Вычисление точности (доля N-грамм из сгенерированного текста, которые присутствуют в эталонных ответах).
Brevity Penalty: Penalization маленьких сгенерированных текстов.
Пример кода для расчета BLEU с использованием библиотеки [nltk:](nltk:)

import nltk
from nltk.translate.bleu_score import corpus_bleu

# Соглашение: мы определяем один или несколько эталонных ответов
reference_corpus = [
    ["это", "чудесный", "день"],
    ["сегодня", "отличная", "погода"]
]
# Генерируемый текст
candidate_corpus = [
    ["это", "прекрасный", "день"],
    ["сегодня", "великолепная", "погода"]
]

# Расчет BLEU
bleu_score = corpus_bleu(reference_corpus, candidate_corpus)
print(f"BLEU score: {bleu_score:.4f}")  # Результат может быть, например, 0.7746
ROUGE Score
ROUGE фокусируется на recall и в основном используется для оценки краткого изложения. Он может измерять:

ROUGE-N: Recall для N-грамм.
ROUGE-L: Длинная последовательность (Longest Common Subsequence).
Пример кода для расчета ROUGE с использованием библиотеки [rouge-score:](rouge-score:)

from rouge_score import rouge_scorer

# Генерируемый текст и эталон
generated_text = "это хороший день"
reference_text = "это прекрасный день"

# Создание объекта оценщика ROUGE
scorer = rouge_scorer.RougeScorer(['rouge1', 'rouge2', 'rougeL'], use_stemmer=True)

# Оценка
scores = scorer.score(reference_text, generated_text)
print(f"ROUGE-1: {scores['rouge1']}")
print(f"ROUGE-2: {scores['rouge2']}")
print(f"ROUGE-L: {scores['rougeL']}")
Приемлемые значения

BLEU Score: Значения выше 0.5 считаются хорошими для большинства приложений, хотя для более специфичных задач высшие значения лучше. Значение 0.7–0.75 может считаться отличным для различных задач NLP.

ROUGE Score: ROUGE-с scores не стандартизированы, но в целом, чем выше, тем лучше. Значения ROUGE-R ниже 0.5 могут сигнализировать о низком качестве, в то время как значения от 0.6 и выше обычно считаются хорошими.

Perplexity
Перплексия (Perplexity) — это метрика, используемая для оценки языковых моделей, которая измеряет, насколько хорошо модель предсказывает последовательности слов. Это важный показатель качества для статистических моделей, включая языковые модели на основе глубокого обучения, такие как LLM (Large Language Models). Перплексия можно рассматривать как обобщение меры “количества удивления” модели, когда она пытается предсказать следующее слово в последовательности. Это вычисляется как экспонента отрицательной логарифмической вероятности, присвоенной истинной последовательности слов.

Перплексия может быть интерпретирована как среднее “число возможных следующих слов” для модели. Чем ниже значение перплексии, тем лучше модель. Это говорит о том, что модель с большей вероятностью предсказала правильные слова.

Чем ниже значение перплексии, тем лучше качество модели. Например, перплексия 20 означает, что модель рассматривает 20 равновероятных следующих слов.
Высокие значения, например, более 100, могут указывать на то, что модель неэффективно предсказывает последовательности.
Пример кода, который рассчитывает перплексию для языковой модели на основе токенов. Будем использовать библиотеку transformers от Hugging Face для выполнения вычислений.
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM # Загружаем модель и токенизатор
model_name = "gpt2"  # Заменить на нужную модель
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name) # Функция для вычисления перплексии
def calculate_perplexity(text):
    # Токенизация текста
    inputs = tokenizer(text, return_tensors="pt")
    # Получаем логиты от модели
    with torch.no_grad():
        outputs = model(inputs, labels=inputs["input_ids"])
        log_likelihood = outputs.loss.item()  # Получаем значение потерь (loss)
    # Перплексия
    perplexity = torch.exp(torch.tensor(log_likelihood))
    return perplexity.item() # Пример текста
text = "Пример текста для расчета перплексии."
perplexity = calculate_perplexity(text)
print(f"Perplexity: {perplexity:.2f}")
Для расчета перплексии с использованием модели Mistral, можно использовать библиотеку Hugging Face Transformers, если Mistral доступен в этой библиотеке.

pip install transformers torch

import torch
from transformers import AutoTokenizer, AutoModelForCausalLM

# Загружаем модель Mistral и токенизатор
model_name = "mistral/mistral"  # Актуальное имя модели
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Убедитесь, что модель в режиме оценки
model.eval()

# Функция для вычисления перплексии
def calculate_perplexity(text):
    # Токенизация текста
    inputs = tokenizer(text, return_tensors="pt")
    
    # Получаем логиты от модели
    with torch.no_grad():
        # Прогоняем входы через модель
        outputs = model(**inputs, labels=inputs["input_ids"])
        log_likelihood = outputs.loss.item()  # Получаем значение потерь (loss)
    
    # Расчет перплексии
    perplexity = torch.exp(torch.tensor(log_likelihood))
    return perplexity.item()

# Пример текста
text = "Это пример текста для расчета перплексии." # Тексты для расчета перплексии можно подставлять разные, чтобы проверить производительность модели на различных входах
perplexity = calculate_perplexity(text)
print(f"Perplexity: {perplexity:.2f}")
b. Человеческая оценка
Использование опросов для оценки ответов модели. Можно использовать платформы для краудсорсинга, такие как Amazon Mechanical Turk, чтобы собрать мнения от людей.

2. Использование автотестов
import requests
import json
from sklearn.metrics import accuracy_score

# Импортируем тестовые данные
with open('test_data.json') as f:
    test_data = json.load(f)

true_answers = []
predicted_answers = []

for item in test_data:
    question = item['question']
    expected_answer = item['expected_answer']
    
    # Запрос к API Mistral
    response = requests.post("http://your-mistral-api-endpoint", json={"input": question})
    answer = response.json().get("output")

    true_answers.append(expected_answer)
    predicted_answers.append(answer)

# Оценка качества
accuracy = accuracy_score(true_answers, predicted_answers)
print(f"Accuracy: {accuracy:.2f}")
Использование unittest, pytest или любой другой фреймворк для тестирования может сделать процесс автотестирования с использованием кода, написанного выше, более структурированным. Пример с использованием [unittest:](unittest:)

import unittest
import requests
import json

class TestMistralResponses(unittest.TestCase):

    def setUp(self):
        with open('test_data.json') as f:
            self.test_data = json.load(f)

    def test_answers(self):
        for item in self.test_data:
            question = item['question']
            expected_answer = item['expected_answer']
            response = requests.post("http://your-mistral-api-endpoint", json={"input": question})
            answer = response.json().get("output")
            self.assertEqual(answer, expected_answer, f"Expected {expected_answer} for question {question}, got {answer}")

if __name__ == "__main__":
    unittest.main()
3. Использование инструментов для оценки качества работы
Weights & Biases (W&B)
Weights & Biases предоставляет платформу для отслеживания экспериментов, визуализации, мониторинга и совместной работы.

pip install wandb

import wandb
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

# Инициализация W&B
wandb.init(project="mistral-evaluation", name="example-evaluation") # Загружаем модель и токенизатор Mistral
model_name = "mistral/mistral"  # Имя модели
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)
# Пример текста
text = "Это пример текста для генерации."
# Токенизация и генерация
inputs = tokenizer(text, return_tensors="pt")
with torch.no_grad():
    outputs = model.generate(inputs, max_length=50)generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
# Логирование результатов в W&B
wandb.log({"input_text": text, "generated_text": generated_text})
# Завершение сессии W&B
wandb.finish()
Hugging Face’s evaluate
Hugging Face предоставляет библиотеку для оценки модели, которая поддерживает множество метрик. Установка:
pip install evaluate
import evaluate
from transformers import AutoModelForCausalLM, AutoTokenizer
import torch

# Загружаем модель и токенизатор Mistral
model_name = "mistral/mistral"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)
# Пример текста
text = "Это пример текста для генерации."
# Токенизация и генерация
inputs = tokenizer(text, return_tensors="pt")
with torch.no_grad():
    outputs = model.generate(inputs, max_length=50)generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
# Подсчет BLEU-метрики
bleu = evaluate.load("bleu")
results = bleu.compute(predictions=[generated_text], references=[[text]])

print("BLEU Score:", results["bleu"])
OpenAI Evaluation Framework
pip install openai

import openai
from transformers import AutoModelForCausalLM, AutoTokenizer
# Загружаем модель и токенизатор Mistral
model_name = "mistral/mistral"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)
# Настройка OpenAI API
openai.api_key = "your-openai-api-key"
# Пример генерации
text = "Это пример текста для генерации."
inputs = tokenizer(text, return_tensors="pt")
with torch.no_grad():
    outputs = model.generate(inputs, max_length=50)generated_text = tokenizer.decode(outputs[0], skip_special_tokens=True)
# Оценка с использованием OpenAI
response = openai.ChatCompletion.create(
    model="gpt-3.5-turbo",  # Можно заменить на актуальную модель OpenAI
    messages=[
        {"role": "user", "content": f"Оцени следующий текст: {generated_text}"}
    ]
)

print("OpenAI evaluation:", response['choices'][0]['message']['content'])
